src/Microsoft.AspNetCore.SignalR.Core/HubConnectionContext.cs

```diff
@@ -28,24 +28,29 @@ public class HubConnectionContext
         private readonly ILogger _logger;
         private readonly CancellationTokenSource _connectionAbortedTokenSource = new CancellationTokenSource();
         private readonly TaskCompletionSource<object> _abortCompletedTcs = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
-        private readonly long _keepAliveDuration;
+        private readonly TimeSpan _keepAliveInterval;
+        private readonly TimeSpan _clientTimeoutInterval;
         private readonly SemaphoreSlim _writeLock = new SemaphoreSlim(1);
 
-        private long _lastSendTimestamp = Stopwatch.GetTimestamp();
+        private long _lastSendTimestamp = DateTime.UtcNow.Ticks;
+        private long _lastReceivedTimestamp = DateTime.UtcNow.Ticks;
         private ReadOnlyMemory<byte> _cachedPingMessage;
+        private bool _clientTimeoutActive;
 
         /// <summary>
         /// Initializes a new instance of the <see cref="HubConnectionContext"/> class.
         /// </summary>
         /// <param name="connectionContext">The underlying <see cref="ConnectionContext"/>.</param>
-        /// <param name="keepAliveInterval">The keep alive interval.</param>
+        /// <param name="keepAliveInterval">How often this client is pinged.</param>
         /// <param name="loggerFactory">The logger factory.</param>
-        public HubConnectionContext(ConnectionContext connectionContext, TimeSpan keepAliveInterval, ILoggerFactory loggerFactory)
+        /// <param name="clientTimeoutInterval">Clients we haven't heard from in this interval are assumed to have disconnected.</param>
+        public HubConnectionContext(ConnectionContext connectionContext, TimeSpan keepAliveInterval, ILoggerFactory loggerFactory, TimeSpan? clientTimeoutInterval = null)
```

Comment:
Changing the public ctor is an API breaking change. What's the policy here @anurse ?

Reply:
Ah yeah, we're going to need to talk about breaking changes :).